### PR TITLE
DFT overhaul, ECPs, SAPT(DFT), EP2, performance enhancements

### LINKED ISSUE #39: Re-enable SCS double hybrids
This seems a little silly that we can't do this at present (after a recent upgrade).

### LINKED ISSUE #40: Make properties work for something besides CC
check other methods

### PR DESCRIPTION
**Pull request by dgasmith**
# Description
Substantial changes to the Psi4 DFT infrastructure to add several hundred new DFT functional forms. In addition, adds the SAPT(DFT) and EP2 methods. Several major performance enhancements are included as well as well. This is also one of the primary patches for the Psi4 1.2 release.

## Todos
Notable points that this PR has either accomplished or will accomplish.
* **Developer Interest**
  - [x] SCF Wavefunctions can now do Hessian-Vector builds see `RHF.Hx` and `RHF.solve` for examples. This functionality supports non-totally symmetric tensors.
  - [x] The DFT V Potential object now has Hessian-Vector functionality.
  - [x] Added many C++ calls to the Python interface.
  - [x] Reworked the building of superfunctionals. Note that the coefficients set when building a functional are no longer modified by exact exchange or MP2 correlation factors.
  - [x] DFT Functionals now takes advantage of polarized vs unpolarized DFT kernels.
  - [x] Added generalized solvers in `p4util/solvers.py`. These include DIIS, CG, BiCGSTAB, and Davidson.
  - [x] New density-fitted AO->MO transform object. More efficient than previous objects and has multiple modes depending on the type of computation (direct or AO storage).
  - [x] ECP SCF energies.
  - [x] ECP gradients by finite differences in Python.
  - [x] Test case(s) and documentation.
  - [x] Make sure frozen core code doesn't add contributions from ECP bearing atoms.
  - [x] Add / document more basis sets, and add ECP info to existing Def2 series (w/help from @loriab).

* **User-Facing for Release Notes**
  - [x] Psi4's DFT now uses LibXC, several hundred new functionals are available.
  - [x] DFT potential and gradient builds are now parallelized and vectorized, up to 30x faster on six cores.
  - [x] Adds a new vectorized DF-JK builder, up to 4x as fast as previous version.
  - [x] Adds culoumb-attenuated (CAM) functionals.
  - [x] Adds the VV10 non-local correlation kernel.
  - [x] Adds DSD functionals.
  - [x] Extended second-order SCF convergence to DFT Wavefunctions.
  - [x] Adds the gradient regularized asymptotic correction (GRAC) to DFT through the option `DFT_GRAC_SHIFT`.
  - [x] Adds a density-fitted second-order Electron Propagator code. 
  - [x] Adds the density-fitted SAPT(DFT) code.
  - [x] Adds linear response DFT.
  - [ ] Adds TDDFT methodologies.
  - [x] To avoid conflict the function `property` was renamed to `properties`.
  - [x] More involved JSON input/output interface.
  - [x] Effective Core Potentials available for SCF energies and gradients, as well as post-SCF energies.
  - [x] Def2 family of basis sets available for all atoms supported by those basis sets.

* **Bug Fixes**
  - [x] wB97X had an incorrect range-seperation parameter. This is now fixed. Note that this did *not* effect wB97X-D.
  - [x] Memory limitations are better adhered to for very large MCSCF computations.

* **TODO**
  - [ ] Run `sed` commands from https://github.com/psi4/psi4/pull/690
  - [x] Pull in ECPs #681 
  - [ ] Rebase in current master to resolve conflicts
  - [x] Closes #39 
  - [x] Closes #40 
  - [ ] New Hessian/Frequency analysis code.
  - [ ] Fill out remaining DFT documentations
  - [ ] Polish new DF JK/Transformer technology
  - [ ] Triple check UKS GGA Hessian-vector code
  - [ ] Finished DFT Hessian functionality (skeleton terms)
  - [ ] Rebase in License/1.1 changes
  - [ ] Rebase in ECP changes (#681)
  - [ ] Re-examine VV10 default grid size
  - [ ] Implement VV10 gradients.

## Status
- [ ]  Ready to go

### CODE REVIEW

**Comment by jgonthier** [comment]
Very impressive work... I'm happy SAPT(DFT) is done. I don't have time to check the details of the equations but I'm confident you did it right. I didn't see any problems after a quick overview of the SAPT code, at least.

Is there a specific reason why you chose to rewrite the SAPT JK energy terms (electrostatic, exchange, induction) in Python instead of having a new function call on the C++ side like you did for dispersion ?

**Comment by dgasmith** [comment]
Hi Jerome,
It was mostly a difficulty in getting the DFT orbital-Hessian vector code accessible in the right places. I had planned on improving that section so it could do mixed input Wavefunctions (RKS, UHF, ROHF, etc). Probably will not get around to it now. 

It's also a neat experiment in mixing of high and low level languages in relation to the readability of code. 

-Daniel Smith

**Comment by jgonthier** [comment]
The code seems quite readable to me right now, which is really nice. When I have some time (so... in 2018 ??) I hope I can come back to Psi4 and play with the code again.

File: tests/python/pubchem/CMakeLists.txt [hunk 1]
```diff
@@ -1,3 +1,3 @@
 include(TestingMacros)
 
-add_regression_test(python-pubchem "psi;quicktests;minitests;python")
+add_regression_test(python-pubchem "psi;minitests;python")
```
Comment (loriab) [inline tests/python/pubchem/CMakeLists.txt]: I think `minitests` became `smoketests` at some point.

File: .scripts/travis_run_test.py [hunk 2]
```diff
@@ -44,3 +43,7 @@
 with open('full_ctest_output.dat', 'w') as outfile:
     outfile.write(ctestout)
```
Comment (andysim) [inline .scripts/travis_run_test.py]: this debug info is quite useful on failures, no?

Comment (dgasmith) [inline .scripts/travis_run_test.py]: It can be very useful; however, its often exceedingly long. I switch it on when I cant debug something any other way.

**Pull request merged**